src/finalizer/utils/helios.ts

```
@@ -0,0 +1,607 @@
+    const encodedMessage = ethers.utils.defaultAbiCoder.encode(["address", "bytes"], [l1Target, l1Data]);
+    // ------------------------------------
+
+    const executeArgs = [proof.sourceNonce, encodedMessage, decodedOutputs.newHead]; // Use encodedMessage
+    const executeTx: AugmentedTransaction = {
+      contract: universalSpokePoolContract,
+      chainId: l2ChainId,
+      method: "executeMessage",
+      args: executeArgs,
+      unpermissioned: true,
+      // @dev Simulation of `executeMessage` depends on prior state update via SP1Helios.update
+      canFailInSimulation: true,
+      message: `Finalize Helios msg (HubPoolStore nonce ${proof.sourceNonce.toString()}) - Step 2: Execute on SpokePool`,
+    };
+    transactions.push(executeTx);
+    crossChainMessages.push({
+      type: "misc",
+      miscReason: "ZK bridge finalization (Execute Message)",
+      originationChainId: l1ChainId,
+      destinationChainId: l2ChainId,
+    });
+  }
+
+  const totalFinalizations = successfulProofs.length + needsExecutionOnlyMessages.length;
+  logger.debug({
+    at: `Finalizer#heliosL1toL2Finalizer:generateHeliosTxns:${l2ChainId}`,
+    message: `Generated ${transactions.length} transactions for ${totalFinalizations} finalizations (${successfulProofs.length} full, ${needsExecutionOnlyMessages.length} exec only).`,
+    proofNoncesFinalized: successfulProofs.map((p) => p.sourceNonce.toString()),
+    execOnlyNoncesFinalized: needsExecutionOnlyMessages.map((m) => m.l1Event.args.nonce.toString()),
+  });
+
+  return { callData: transactions, crossChainMessages: crossChainMessages };
+}
+
+/**
+ * Retrieves an ethers.Contract instance for the SP1Helios contract on the specified chain.
+ * @throws {Error} If the SP1Helios contract address or ABI is not found for the given chainId in CONTRACT_ADDRESSES.
+ */
+function getSp1HeliosContract(chainId: number, signerOrProvider: Signer | Provider): ethers.Contract {
```

nit: should we move `getSp1HeliosContract` and  `getHubPoolStoreContract` into `UniversalUtils`?

---

Well. It's nuanced. These fns are opinionated in the sense that they throw if the contract is not there. This throwing opinion is really of `helios.ts`, not of utils package. But I agree that we should probably always throw on missing addr / abi so we can move?

---

Also, I don't feel like these fns are super-useful to reuse. I don't mind either way

---

i don't think the opinionated throwing is a bad thing to have in a utility function named as such. We should move it to a utility folder.